pageserver/src/tenant.rs

```diff
@@ -1639,64 +1644,86 @@ impl Tenant {
-                        activated_timelines += 1;
-                    }
+            for timeline in not_broken_timelines {
+                timeline.activate(broker_client.clone(), ctx);
+                activated_timelines += 1;
+            }
 
-                    let elapsed = self.loading_started_at.elapsed();
-                    let total_timelines = timelines_accessor.len();
+            self.state.send_modify(move |current_state| {
+                assert_eq!(
+                    *current_state,
+                    TenantState::Activating,
+                    "set_stopping and set_broken wait for us to leave Activating state",
+                );
+                *current_state = TenantState::Active;
 
-                    // log a lot of stuff, because some tenants sometimes suffer from user-visible
-                    // times to activate. see https://github.com/neondatabase/neon/issues/4025
-                    info!(
-                        since_creation_millis = elapsed.as_millis(),
-                        tenant_id = %self.tenant_id,
-                        activated_timelines,
-                        total_timelines,
-                        post_state = <&'static str>::from(&*current_state),
-                        "activation attempt finished"
-                    );
-                }
-            }
-        });
-        result
+                let elapsed = self.loading_started_at.elapsed();
+                let total_timelines = timelines_accessor.len();
+
+                // log a lot of stuff, because some tenants sometimes suffer from user-visible
+                // times to activate. see https://github.com/neondatabase/neon/issues/4025
+                info!(
+                    since_creation_millis = elapsed.as_millis(),
+                    tenant_id = %self.tenant_id,
+                    activated_timelines,
+                    total_timelines,
+                    post_state = <&'static str>::from(&*current_state),
+                    "activation attempt finished"
+                );
+            });
+        };
+        Ok(())
     }
 
-    /// Change tenant status to Stopping, to mark that it is being shut down
-    pub fn set_stopping(&self) {
+    /// Change tenant status to Stopping, to mark that it is being shut down.
+    ///
+    /// This function is not cancel-safe!
+    pub async fn set_stopping(&self) {
+        // `Activating` is a transient state during which no external state transitions are supported.
+        let mut rx = self.state.subscribe();
+        rx.wait_for(|state| state != TenantState::Activating)
+            .await
+            .expect("cannot drop self.state while on a &self method");
+
+        let mut stopping = false;
         self.state.send_modify(|current_state| {
             match current_state {
+                TenantState::Activating => unreachable!("we checked above and never transition back into Activating state"),
```

Comment:
Thinking about is this really unreachable?

We wait_for state to not be activating, but I don't see this covering exhaustively.
| t   | first task loading | second task trying set_stopping      |
| --- | ------------------ | ------------------------------------ |
| t=0 | state=Loading      |                                      |
| t=1 |                    | check state!=Activating              |
| t=2 | state==Activating  |                                      |
| t=3 |                    | unreachable!("we checked above...")  |

Or what am I missing? Could be something obvious.

Reply:
Right, need to move the stuff inside the `send_modify` closure into the `wait_for` closure.
Luckily, wait_for passes a mut ref

Reply:
Got an idea

Reply:
My idea was to just loop back if we find ourselves in the current `unreachable` within `send_modify`.

Reply:
Oki I see now my `set_stopping` should had been `set_broken`.

Reply:
Oki, de780d2 looks quite ok to me. I do hope we don't get other ideas of setting tenants to broken than the ones we currently have in tests. Luckily those are `active -> broken`, so enabling that for non-testing builds does not sound like a huge chore.

Resolving.